### DISCUSSION TITLE
milvus 并发搜索效率比 faiss低10倍，why？

### DISCUSSION THREAD

**Discussion by kkeyong** [milvus-io/milvus#16813] [Q&A and General discussion]
图一：faiss并发100效率
![image](https://user-images.githubusercontent.com/20456487/167114185-899a741d-14ad-4971-9694-033b71bac137.png)
图二：miluvs并发100效率
![image](https://user-images.githubusercontent.com/20456487/167114297-21535147-4cf5-4db2-b254-6328457cffc1.png)

hello，milvus单机search效率比faiss低10倍？应该怎么优化？求指教

**Comment by xiaofan-luan** [comment]
请问是什么版本，另外查询参数和索引参数分别是什么？

**Reply by kkeyong** [reply]
版本2.0.2，单机部署。数据量10w，64维
索引参数：{"metric_type":"IP", "index_type":"IVF_FLAT","params":{"nlist":1024}}
查询参数：{"metric_type": "IP", "params": {"nprobe": 64}}
faiss与milvs的参数设置一致
并发时，48核机器只用了差不多5~6个核。不知道是什么限制了。而且并发时，速度越来越慢。

**Reply by xiaofan-luan** [reply]
nq小的场景下，如果并发上不去，一般是因为python客户端的锅， 在2.1会对小nq的场景有一个巨大的性能优化，再有1-2周就会发布了

**Reply by xiaofan-luan** [reply]
建议等我们2.1正式放出来在测试，会有惊喜

**Reply by kkeyong** [reply]
目前版本其他客户端能应对小nq的场景吗？

**Reply by kkeyong** [reply]
> nq小的场景下，如果并发上不去，一般是因为python客户端的锅， 在2.1会对小nq的场景有一个巨大的性能优化，再有1-2周就会发布了

我们尝试了用java，速度是一样的。

**Reply by yhmo** [reply]
10万数据量的查询速度正常应该在十几毫秒这样，你测得的平均每个请求几百毫秒可能是consistency_level设成了Strong。把consistency_level参数设置成Eventually试试
results = collection.search(
	......
	consistency_level="Eventually"
)

由于milvus有多个节点间rpc的消耗，耗时会比faiss高一些，在小数据集的时候差距显得比较明显。
CPU没用满是因为数据量太小，一下就计算完了，就好比汽车刚启动就踩刹车，不会达到最高利用率。一般到千万以上级别数据量就能看到较高的CPU利用率。

**Reply by xiaofan-luan** [reply]
@yhmo python已经把一致性等级调整成bounded了，java可以也调整下

**Reply by yhmo** [reply]
> @yhmo python已经把一致性等级调整成bounded了，java可以也调整下

这个平均500ms的单次请求耗时不像是bounded...

**Reply by kkeyong** [reply]
> consistency_level="Eventually"

1、500ms是通过多线程并发方式请求的耗时，单次请求这边也是十几毫秒。
2、consistency_level没有调整过，应该是默认值
3、我尝试把consistency_level 改成 "Eventually"，结果更慢了。

**Reply by kkeyong** [reply]
> 
每个并发都是nq=1的search

**Reply by xiaofan-luan** [reply]
应该就是并发上去以后请求排队的问题，建议等一下2.1，对这个场景会有大幅性能提升。

**Reply by kkeyong** [reply]
> 应该就是并发上去以后请求排队的问题，建议等一下2.1，对这个场景会有大幅性能提升。

好的。
另外一个问题，为啥使用docker-compose部署的分布式milvus，增加querynode后，没有性能提升。而且部分querynode没有真正执行search操作（没有log和占用CPU）

**Reply by xiaofan-luan** [reply]
> > 应该就是并发上去以后请求排队的问题，建议等一下2.1，对这个场景会有大幅性能提升。
> 
> 好的。 另外一个问题，为啥使用docker-compose部署的分布式milvus，增加querynode后，没有性能提升。而且部分querynode没有真正执行search操作（没有log和占用CPU）

大概有多少数据量呢？一个segment大约500M，数据量太少了话加节点可能没有balance效果

**Reply by ruixiu** [reply]
2.1什么时候发布

**Reply by yhmo** [reply]
> 2.1什么时候发布

2.1分支上周已经分出，正在做最后的冲刺。快了...

**Reply by ruixiu** [reply]
2.0.2遇到了两个无法解决的问题：1.内存占用超高；2.milvus重启后搜索超时（等待2天仍不可用）。迫切等待2.1
